text/0000-nested-publish.md

```diff
@@ -0,0 +1,186 @@
+
+You can change this default by placing in the manifest (`Cargo.toml`) of a sub-package:
+
+```toml
+[package]
+publish = "nested"
+```
+
+If this is done, Cargo's behavior changes as follows:
+
+* If you publish the parent package, the sub-package is included in the `.crate` file (unless overridden by explicit `exclude`/`include`) and will be available to the parent package whenever the parent package is downloaded and compiled.
+* The parent package (and other sub-packages) may have `path =` dependencies upon the sub-package. (Such dependencies must not have a `version =` or `git =`; that is, the `path` must be the _only_ source for the dependency.)
+* You cannot `cargo publish` the sub-package, just as if it had `publish = false`. (This is a safety measure against accidentally publishing the sub-package separately when this is not intended.)
+
+Nested sub-packages may be freely placed within other nested sub-packages.
+
+When a group of packages is published in this way, and depended on, this has a number of useful effects (which are not things that Cargo explicitly implements, just consequences of the system):
+
+* The packages are a single unit for all versioning purposes; there is no way for a version mismatch to arise since all the code was published together. Version resolution does not apply (in the same way that it does not for any other `path =` dependency).
+* The sub-package is effectively “private”: it cannot be named by any other package on `crates.io`, only by its parent package and sibling sub-packages.
+
+## Example: trait and derive macro
+
+Suppose we want to declare a trait-and-derive-macro package. We can do this as follows. The parent package would have this manifest `foo/Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros" }    # newly permitted
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"                     # this name need not be claimed on crates.io
+version = "0.1.0"                   # this version is not used for dependency resolution
+edition = "2021"
+publish = "nested"                  # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent `foo` directory, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following changes must be made across Cargo and `crates.io`:
+
+* **Manifest schema**
+    * The Cargo manifest now allows `"nested"` as a value for the `package.publish` key.
```

Comment:
As it is already a term in `cargo`, I actually lean towards `"vendor"`

Reply:
I get the similarity, but vendoring normally means making a _copy_ of a package that is available by other means, and one of the design goes here is to _discourage_ any such copies existing (because they are likely to be accidental, and if they aren't, then they may create the same kinds of problems as multiple major versions do). I think reusing the term would create more confusion than it avoids.

Reply:
> vendoring normally means making a copy of a package that is available by other means, an

The perspective I was using when I came up with "vendor" was that  instead of getting a dependency through the registry, we are copying it into our package.  Its not vendored within the repo but in the `.crate` file.

This also ties into whether we should generalize this across dependency sources at which point it feels like it becomes even more applicable.

Reply:
I still don't think that vendoring is the right term, _especially_ as one of the things that came up during my review of prior art is the concept of using nested-packages-or-whatever _for_ vendoring, e.g. to fix a bug before upstream accepts the patch — I think these need to be kept distinct ideas.

That being considered, what do we need to do here with the RFC text to resolve this thread? Should there be an unresolved question for terminology, or can we just proceed as-is?

Reply:
I think "vendoring" isn't all-inclusive of the situations where one might use this. Usually vendoring refers to "taking some third party dependency's entire source code and jamming it into to some `vendor/whatever` directory (or sometimes via a git submodule) for _reasons_". I don't think it makes sense to use this term in scenarios where the nested crate _isn't_ third party, which I think will actually be most of the time with this feature

Reply:
I haven't seen any "seconding" of support for the `"vendor"` name, so I'm going to keep `"nested"` and resolve this. (Of course, there might be some third or fourth better idea to be found…)

Reply:
At least keeping this unresolved to centralize any name bikeshedding conversations